R/rma-tidiers.R

```diff
@@ -0,0 +1,288 @@
+#' @templateVar class rma
+#' @template title_desc_tidy
+#' 
+#' @param x An `rma` created by the `metafor` package.
+#' @inheritParams tidy.lm
+#' @param include_studies Logical. Should individual studies be included in the
+#'    output?
+#' @template param_unused_dots
+#' @param measure Measure type. See [metafor::escalc()]
+#'
+#' @evalRd return_tidy(
+#'   study = "The name of the individual study",
+#'   type = "The estimate type  (summary vs individual study)",
+#'   "estimate",
+#'   "std.error",
+#'   "statistic",
+#'   "p.value",
+#'   "conf.low",
+#'   "conf.high"
+#' )
+#' @export
+#'
+#' @examples
+#' 
+#' library(metafor)
+#' 
+#' df <-
+#'   escalc(
+#'     measure = "RR",
+#'     ai = tpos,
+#'     bi = tneg,
+#'     ci = cpos,
+#'     di = cneg,
+#'     data = dat.bcg
+#'   )
+#' 
+#' meta_analysis <- rma(yi, vi, data = df, method = "EB")
+#'
+#' tidy(meta_analysis)
+#'
+#' @rdname metafor_tidiers
+#' 
+tidy.rma <- function(x, conf.int = FALSE, conf.level = 0.95, exponentiate = FALSE,
+                     include_studies = TRUE, measure = "GEN", ...) {
+  # tidy summary estimates
+  betas <- x$beta
+  if (!is.null(nrow(betas)) && nrow(betas) > 1) {
+    # get estimate type and fix spelling
+    study <- rownames(betas)
+    swap <- grepl("intrcpt", study)
+    study[swap] <- "intercept"
+    betas <- as.double(betas)
+  } else {
+    study <- "overall"
+    betas <- betas[1]
+  }
+  
+  results <- tibble::tibble(
+    study = study, 
+    type = "summary",
+    estimate = betas, 
+    std.error = x$se,
+    statistic = x$zval, 
+    p.value = x$pval,
+    conf.low = x$ci.lb, 
+    conf.high = x$ci.ub
+  )
+  
+  # tidy individual studies
+  if (include_studies) {
+    # use `metafor::escalc` to standardize estimates and confidence intervals
+    estimates <- metafor::escalc(yi = x$yi.f, vi = x$vi.f, measure = measure) %>%
+      summary(level = conf.level * 100) %>% 
+      as.data.frame(stringsAsFactors = FALSE)
+    
+    n_studies <- length(x$slab)
+    
+    estimates <- dplyr::bind_cols(
+      study = as.character(x$slab), 
+      # dplyr::bind_cols is strict about recycling, so repeat for each study
+      type = rep("study", n_studies), 
+      estimates[, c("yi", "sei", "zi")], 
+      p.value = rep(NA, n_studies),
+      estimates[, c("ci.lb", "ci.ub")] 
+    )
+    
+    names(estimates) <- c("study", "type", "estimate", "std.error", "statistic",
+                          "p.value", "conf.low", "conf.high")
+    estimates <- tibble::as_tibble(estimates)
+    results <- dplyr::bind_rows(estimates, results) 
+  }
+  
+  if (exponentiate) {
+    results <-  dplyr::mutate_at(results, dplyr::vars(estimate, conf.low, conf.high), exp)
+  }
+  
+  if (!conf.int) {
+    results <- dplyr::select(results, -conf.low, -conf.high)
+  }
+  
+  # remove extra model data from study names
+  attributes(results$study) <- NULL
```

Comment:
Why is this necessary?

Reply:
Hm, metafor stores a lot of extra information in these vectors which caused problems in working with data frames, but I suspect being stricter while assembling `tibbles` and using `dplyr::bind_*()` already stripped these and now it's no longer necessary.